### ISSUE TITLE
Add more events to CollectionView items: long tap, swipe right

### ISSUE THREAD

**Issue by sand123** [eclipsesource/tabris-js#567]
Hi!
It would be great to add such events to CollectionView: long tap for opening context menu, swipe right fo removing item from list, for example

**Comment by wilsonna** [comment]
@sand123 you can already do this with the current API. Below is my code for your reference. In both cases, "composite" is a Composite widget appended in a Cell of the CollectionView. You can also refer to the documentation on Touch Events (https://tabrisjs.com/documentation/latest/touch) for a list of available events you can use to implement what you need.

**Longpress**

    composite.on("longpress", function(widget, gesture) {
        if (gesture.state === "start") {
            // do what you want here
        }
    });

**Slide to Delete**
This allows you slide left or right. If you slide more than 100 pixels, it will slide the composite out of the screen and show a confirm or undo button. Anything less than 100 pixels, it will revert back to its original position.

    composite.on("pan:horizontal", function(this, gesture) { // Remove item
        this.set("transform", {translationX:gesture.translation.x});
        if (gesture.state === "end") {
            if (Math.abs(gesture.translation.x) > 100) {
                this.animate({transform:{translationX:(gesture.translation.x<0 ? -1 : 1)*1000}},{duration:1000});
                var confirmationComposite = tabris.create("Composite", {
                    layoutData: {centerY:0, centerX:0}
                }).appendTo(this.parent());
                tabris.create("Button", {
                    layoutData: {centerY:0, left:confirmationComposite.children().last()},
                    text: "Remove"
                }).on("select", function() {
                    // do what you want here
                }).appendTo(confirmationComposite);
                tabris.create("Button", {
                    layoutData: {centerY:0, left:confirmationComposite.children().last()},
                    text: "Undo"
                }).on("select", function() {
                    this.animate({transform:{translationX:0}},{duration:1000});
                }).appendTo(confirmationComposite);
            } else {
                this.animate({transform:{translationX:0}},{duration:1000});
            }
        }
    });

**Comment by sand123** [comment]
@wilsonna Ok, as a workaround. But
1) for the performance reasons, I think it's better to delegate items events (just like `select`) to CollectionView itself, rather then bind same listeners to every element. 
2) for simple list, I don't want to wrap cell content into Composite just to have these events fired - they should not depend on cell template

**Comment by ralfstx** [comment]
@sand123 I don't think that @wilsonna's solutions are workarounds, that's the recommended way to attach listeners, see the [swipe-to-dismiss example](https://github.com/eclipsesource/tabris-js/tree/master/examples/swipe-to-dismiss).

1) Why do you think that delegating cell events to the CollectionView could improve performance? Cell is the widget that actually receives the event.
2) You can add the listener to the Cell itself, there's no need for wrapping.

If this API doesn't make sense for you, could you draft an example snippet of what you'd have expected?

**Comment by wilsonna** [comment]
@sand123 for the sliding event, we may need to have different implementations under different scenarios. For example, instead of sliding to delete, I may instead wish to slide to reveal more information underneath the composite. Furthermore, even for delete, what we want actually do when "slide-out" also varies based on our needs.

What I did to cut down on my boiler plate code is to create a function (Deleter.register) which encompasses the code I wrote above. I then use it to register a "deleter", with a callback function. e.g.

	var collectionView = tabris.create("CollectionView", {
		layoutData: {top:headerComposite, bottom:0, left:0, right:0},
		refreshEnabled: true,
		itemHeight: Global.itemHeight,
		initializeCell: function(cell) {
			var composite = tabris.create("Composite", {
				layoutData: {top:0, bottom:0, left:0, right:0}
			});

			Deleter.register(composite, function(widget) {
				var cell = widget.parent();
				var session = cell.get("item");
				Service.deleteSession(session.id, function() {
					console.log("Session " + session.id + " deleted");
					collectionView.remove(cell.get("itemIndex"));
				});
			});
                        ....
		}
	});

This way, I can reuse my Deleter under various scenarios and implementations

**Comment by wilsonna** [comment]
Regarding @ralfstx second point about attaching events to the Cell instead of a wrapper (i.e. another composite), I actually did it for a not-so-significant reason. It's because I've also appended another composite (not shown in my code above) to the Cell to serve as a row divider. If I were to attach the "pan" event to the Cell, sliding the cell would also slide the divider out of the screen. As such, I had to create a wrapper so that only the wrapper would slide, but the divider would stay. :smile:
